### ISSUE TITLE
query_posts() alternative for pagination with Timber?

### ISSUE THREAD

**Issue by ashmatadeen** [timber/timber#841]
The [wiki on pagination](https://github.com/jarednova/timber/wiki/Pagination) recommends the use of `query_posts()` to get pagination working on page templates.

The [WordPress Codex](https://codex.wordpress.org/Function_Reference/query_posts) on `query_posts` discourages the use of this function. Is there a way of getting pagination working with Timber without `query_posts`?

**Comment by jarednova** [comment]
@ashmatadeen you only need to use `query_posts` when you're using `get_pagination` for something other than the default query.

As for why vs. WP's recommendation: basically `query_posts` slaps WordPress across the face with a new query which can have implications for things like pagination ... the very pagination we're looking to exploit!

It's very possible that as WP has been doing better OO organization there's now (or will be) a better way to grab data when assembling the page links. The problem is, that (at time of writing those functions) the links were heavily dependent on nasty globals and such — which are the globals that `query_posts` is muscling into

**Comment by jarednova** [comment]
... if you can see a better way to get pagination going without using `query_posts` I'd love to merge it. Again, it's very possible that since I wrote this part (prob. around WP 3.8) things in the internals have gotten better

**Comment by ashmatadeen** [comment]
Thanks for the explanation @jarednova I'll post an update here if I find a way around it.

**Comment by ashmatadeen** [comment]
From what I can see, `Timber::get_pagination()` is relying on WP's `query_posts()` for [two properties](https://github.com/jarednova/timber/blob/eb1267ce586747abf47e2568fe10c9fe8fb71865/lib/timber.php#L502):
1. `$wp_query->found_posts`
2. `$wp_query->query_vars['posts_per_page']`

Before running `Timber::get_pagination()` however, Timber has [access](https://github.com/jarednova/timber/blob/eb1267ce586747abf47e2568fe10c9fe8fb71865/lib/timber-query-iterator.php#L57) to both properties when running `Timber::get_posts($args)`. The properties are accessible in `TimberQueryIterator->get_posts()` as:
`$this->_query->found_posts` and `$this->_query->query_vars['posts_per_page']`.

Not sure if this is the best approach but I'm thinking if Timber can keep hold of these two properties to later pass them onto `Timber::get_pagination()` then `Timber::get_pagination()` does not have to rely on those being set in global `$wp_query`? Do you have any insights into how to best achieve this?

**Comment by hyperbrave** [comment]
how do you use a post class for posts queried with query_posts?

**Comment by toddholmberg** [comment]
@jarednova 
Is it possible to do as @ashmatadeen suggests: add found_posts to the pagination object results? I would also like to extend the suggestion, and make found_posts available to Timber post query output.

**Comment by justono1** [comment]
So here is what I did just to get around having to user query_posts to return pagination on a custom query. I had a class that was extending WP_Query to add location based sorting options based on meta_values so I had to use WP_query posts.

I just went into @jarednova Pagination.php in the plugin and saw that in the constructor there is a $wp_query object set to null. If null it just uses the `global $wp_query` object. So I just created a new instance of the Pagination class in my template file and passed it the blank array for the first param and my custom query object as the second everything worked out.

`$pagination = 'Timber\Pagination';`
`$context['pagination'] = new $pagination(array(), $my_query_object);`
